### PR TITLE
Add a linux archive to distribute atom

### PR DESCRIPTION
**Pull request by didrocks**
Now that Atom has a relocable/portable config directories and different channels, it makes sense to have a way to distribute an archive on Linux (as suggested on bug #7102) containing latest available code for each channels.

This has two advantages:
- this will enable (in a later step) to enable auto-upgrade to follow multiple channels (optionally in parallel in different directories). We can even think giving a way for people to switch between channels as needed (given the user config stay backward compatible). I'm happy to work on those 2 problematics after/if this PR is merged at some point.
- if all dependencies are installed in the distribution, the atom installation is then just a question of extracting an archive, requiring non root permission contrary to the .deb or .rpm packages.

We can then integrate this archive to be downloaded to some tools like Ubuntu Make (https://github.com/ubuntu/ubuntu-make/issues/6).

We generate a .tar.gz compressed archive (we could switch to a more performant .tar.xz, but the first format is more widespread and usual). I did add the `mktar` to the `CI task` (I saw the rpm wasn't in there, so I guess, it's only running on an Ubuntu machine) and tested it. I did add the task as well to `upload-assets`, but of course, I couldn't test it ;) The changes there are though minimal, but I can modify what's necessary as needed.

As _version_ contains _upstream version + channel name_, I didn't change anything in the naming of the archive publication, and the folder in archive contains thus an uniquely defined build.

Of course, the resulting archive is linked against the libraries available in the builder version. mktar is run where the .deb is built, which is, I guess, the latest ubuntu LTS. I tried the built archive on 14.04 LTS, 15.04, 15.10 and xenial (dev version), and everything is working as expected. If this doesn't work on other distros (which, looking at the deps, seems unlikely and we already only have one deb for debian/ubuntu and one rpm for fedora/opensuse…), we can build multiple archives. But I propose we starts with this one.

I'm happy to do any changes that you feel are necessary!

### CODE REVIEW

**Comment by didrocks** [comment]
I looked at the check failure: it seems that the failure on "ATOM_SPECS_TASK=packages NODE_VERSION=0.12" darwin config seems to be a network transiant issue (npm returning ECONNRESET). The other configs are passing (and you can see that the `mktar` task is called on Linux machines). I can't relaunch the Travis CI task as only the repo owner could.

**Comment by 50Wliu** [comment]
Restarted it :).

**Comment by didrocks** [comment]
Thanks @50Wliu! Better now :)

File: build/tasks/publish-build-task.coffee [hunk 1]
```diff
@@ -103,10 +103,17 @@ getAssets = ->
...
 
       cp sourcePath, path.join(buildDir, assetName)
+      assets = [{assetName, sourcePath}]
 
-      [
-        {assetName, sourcePath}
-      ]
+      # Check for an archive build on a debian build machine.
+      # We could provide a Fedora one if necessary and not compatible
+      sourcePath = "#{buildDir}/#{appFileName}-#{version}-#{arch}.tar.gz"
```
Comment (50Wliu) [inline build/tasks/publish-build-task.coffee]: The `/` should probably be replaced with `path.join`.

File: build/tasks/publish-build-task.coffee [hunk 2]
```diff
@@ -103,10 +103,17 @@ getAssets = ->
         assetName = "atom.#{arch}.rpm"
 
       cp sourcePath, path.join(buildDir, assetName)
+      assets = [{assetName, sourcePath}]
 
-      [
-        {assetName, sourcePath}
-      ]
+      # Check for an archive build on a debian build machine.
+      # We could provide a Fedora one if necessary and not compatible
```
Comment (50Wliu) [inline build/tasks/publish-build-task.coffee]: Could you elaborate on this comment?  The last part (`and not compatible`) is confusing me a bit.

File: README.md [hunk 3]
```diff
@@ -73,6 +73,22 @@ Currently only a 64-bit version is available.
 The Linux version does not currently automatically update so you will need to
 repeat these steps to upgrade to future releases.
 
+### Archive extraction
+
+An archive is available for people who don't want to install `atom` as root.
+
+This version enables you to install multiple atom versions in parallel. It has been built on Ubuntu 64-bits,
```
Comment (50Wliu) [inline README.md]: `atom` should be capitalized, and `64-bits` should be changed to `64-bit` to be consistent with the rest of the documentation.

File: README.md [hunk 4]
```diff
@@ -73,6 +73,22 @@ Currently only a 64-bit version is available.
...
+but should be compatible with other Linux distributions.
+
+1. Install dependencies (on Ubuntu): `sudo apt install git gconf2 gconf-service libgtk2.0-0 libudev1 libgcrypt20
+libnotify4 libxtst6 libnss3 python gvfs-bin xdg-utils libcap2`
+2. Download `atom-amd64.tar.gz` from the [Atom releases page](https://github.com/atom/atom/releases/latest).
+3. Run `tar xf atom-amd64.tar.gz` in directory you want to extract atom folder.
```
Comment (50Wliu) [inline README.md]: Shouldn't this be `tar -xf ...`?

File: build/tasks/mktar-task.coffee [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+    iconPath = path.join(shellAppDir, 'resources', 'app.asar.unpacked', 'resources', 'atom.png')
+
+    cmd = path.join('script', 'mktar')
+    args = [appFileName, version, arch, iconPath, buildDir]
+    spawn {cmd, args}, (error) ->
+      if error?
+        done(error)
+      else
+        grunt.log.ok "Created #{buildDir}/#{appFileName}-#{version}-#{arch}.tar.gz"
```
Comment (50Wliu) [inline build/tasks/mktar-task.coffee]: Again, `path.join` should be used here.

File: docs/build-instructions/linux.md [hunk 6]
```diff
@@ -74,7 +74,7 @@ If you have problems with permissions don't forget to prefix with `sudo`
 
   To use the newly installed Atom, quit and restart all running Atom instances.
 
-5. *Optionally*, you may generate distributable packages of Atom at `$TMPDIR/atom-build`. Currently, `.deb` and `.rpm` package types are supported. To create a `.deb` package run:
+5. *Optionally*, you may generate distributable packages of Atom at `$TMPDIR/atom-build`. Currently, `.deb` and `.rpm` package types are supported as well as `tar.gz` archive. To create a `.deb` package run:
```
Comment (50Wliu) [inline docs/build-instructions/linux.md]: `as well as a ...`

File: README.md (hunk 4 above)
Comment (paulcbetts) [inline README.md]: Nope, works without it too

Comment (50Wliu) [inline README.md]: Ah, that's nice to know.  Thanks.

File: build/tasks/mktar-task.coffee (hunk 5 above)
Comment (didrocks) [inline build/tasks/mktar-task.coffee]: Will fix! build/tasks/mkdeb-task.coffee didn't use that (and that file was based on it), fixing.

File: README.md (hunk 3 above)
Comment (didrocks) [inline README.md]: changed!

File: README.md (hunk 4 above)
Comment (didrocks) [inline README.md]: yeah, real usage of tar should be without "-" (Unix compatible versions)

File: build/tasks/publish-build-task.coffee (hunk 2 above)
Comment (didrocks) [inline build/tasks/publish-build-task.coffee]: I replaced it with "we could provide a Fedora version if some libraries are not compatible"

**Comment by didrocks** [comment]
Thanks for the review/ Just pushed a new commit fixing the comments.

I did rerun a `script/grunt mktar` after a new build and a `script/grunt ci` to ensure both direct build and CI are still passing as expected.

On travis CI, the 2 failures seems some transient issues:
```
 >> Failed to download electron v0.34.5 [Error: Could not find the End of Central Directory Record]
```
and:
```
>> Error Downloading Update: Could not get code signature for running application
```

Mind relaunching the tests?
Thanks again!

**Comment by 50Wliu** [comment]
The failing core spec actually seems to be my fault...oops.  I'll go revert that change until I can fix the specs.

**Comment by didrocks** [comment]
looks indeed better now :)

File: README.md (hunk 4 above)
Comment (50Wliu) [inline README.md]: Something else I noticed on this line: it should probably be `in the directory where you want to extract the Atom folder`.

File: build/tasks/publish-build-task.coffee [hunk 7]
```diff
@@ -103,10 +103,17 @@ getAssets = ->
...
 
       cp sourcePath, path.join(buildDir, assetName)
+      assets = [{assetName, sourcePath}]
 
-      [
-        {assetName, sourcePath}
-      ]
+      # Check for an archive build on a debian build machine.
+      # We could provide a Fedora version if some libraries are not compatible
+      sourcePath = "#{buildDir}/#{appFileName}-#{version}-#{arch}.tar.gz"
```
Comment (50Wliu) [inline build/tasks/publish-build-task.coffee]: This also needs to use `path.join` :wink:.

File: docs/build-instructions/linux.md [hunk 8]
```diff
@@ -74,7 +74,7 @@ If you have problems with permissions don't forget to prefix with `sudo`
 
   To use the newly installed Atom, quit and restart all running Atom instances.
 
-5. *Optionally*, you may generate distributable packages of Atom at `$TMPDIR/atom-build`. Currently, `.deb` and `.rpm` package types are supported. To create a `.deb` package run:
+5. *Optionally*, you may generate distributable packages of Atom at `$TMPDIR/atom-build`. Currently, `.deb` and `.rpm` package types are supported, as well as `tar.gz` archive. To create a `.deb` package run:
```
Comment (50Wliu) [inline docs/build-instructions/linux.md]: `...as well as a tar.gz...`

File: docs/build-instructions/linux.md [hunk 9]
```diff
@@ -86,6 +86,12 @@ If you have problems with permissions don't forget to prefix with `sudo`
   script/grunt mkrpm
   ```
 
+  To create an `.tar.gz` package run
```
Comment (50Wliu) [inline docs/build-instructions/linux.md]: I don't think the `an` is needed here - just `a`

**Comment by 50Wliu** [comment]
Looks good to me other than a few little nits, but I'm probably not the best reviewer for this :P.
/cc @atom/feedback

File: build/tasks/publish-build-task.coffee (hunk 7 above)
Comment (didrocks) [inline build/tasks/publish-build-task.coffee]: argh, missed that one (mirroring other publishing task, I'll fix the one that are already upstream in a followup PR I guess ;))

**Comment by didrocks** [comment]
And here we go, just push the additional typo/path.join fixes. Thanks for the review!
To ensure I didn't introduce any typo, I've reran build, mktar and ci tasks, and they all pass (as well as Travis CI)

**Comment by yieldman** [comment]
Why aren't we merging this pull request already ?

**Comment by 50Wliu** [comment]
> Why aren't we merging this pull request already ?

Someone with more knowledge in this area of the codebase needs to review it first.  I only did a very basic review.

**Comment by yieldman** [comment]
Could you please refer this PR to appropriate maintainer after New Year ?

**Comment by 50Wliu** [comment]
> Could you please refer this PR to appropriate maintainer after New Year ?

I'll try :smile:.

**Comment by Sisyphuss** [comment]
Waiting for the Linux Zip to use it as a Julia IDE in my laboratory.

**Comment by joshaber** [comment]
I don't feel terribly qualified to review this. @atom/feedback Anyone else?

**Comment by belaustegui** [comment]
Is there any news on this issue? As an Ubuntu user being able to install Atom and update using Ubuntu Make would be incredible useful.

**Comment by didrocks** [comment]
I would be happy to fix the conflicts with trunk if I'm ensured that upstream will look at it this time.

**Comment by lee-dohm** [comment]
@didrocks, let me talk to the team and see what the status is.

**Comment by iolsen** [comment]
@didrocks this is great and we'd be psyched to merge it if you'll fix the conflicts! 📦 ✨

**Comment by didrocks** [comment]
Here is a rebased and squashed version of this PR! The tests on Linux (impacted by the change) are passing. The failure on one Mac machine is already happening on master and thus, doesn't seem related.

**Comment by lee-dohm** [comment]
Thanks @didrocks! I've notified the right people :grinning:

**Comment by thomasjo** [comment]
> The tests on Linux (impacted by the change) are passing. The failure on one Mac machine is already happening on master and thus, doesn't seem related.

Correct assumption. I restarted the failing Travis build, and it's :green_heart:

**Comment by didrocks** [comment]
Thanks a lot for merging! I see though that yesterday's release don't publish the linux tarball yet: https://github.com/atom/atom/releases/tag/v1.7.0. Do you have a pointer to the code for the tool publishing this? I'm happy to patch this as well.

**Comment by lee-dohm** [comment]
@iolsen: Can you help with this too?

**Comment by iolsen** [comment]
Hey @didrocks sorry about that. Getting this integrated into our build/release scripts is going to be difficult because they're kind of a mess right now. Getting that sorted out is a priority but is going to be a blocker for this until it's done. I appreciate your offer to help and I'll take you up on that when I can!

**Pull request merged**